src/inmanta/server/services/environmentservice.py

```diff
@@ -471,21 +482,31 @@ async def environment_list(self, details: bool = False) -> list[model.Environmen
 
     @handle(methods_v2.environment_delete, environment_id="id")
     async def environment_delete(self, environment_id: uuid.UUID) -> None:
-        env = await data.Environment.get_by_id(environment_id)
-        if env is None:
-            raise NotFound("The environment with given id does not exist.")
+        async with self.environment_deletion_lock:
+            async with data.Environment.get_connection() as connection:
+                env = await data.Environment.get_by_id(environment_id, connection=connection)
+                if env is None:
+                    raise NotFound("The environment with given id does not exist.")
 
-        is_protected_environment = await env.get(data.PROTECTED_ENVIRONMENT)
-        if is_protected_environment:
-            raise Forbidden(f"Environment {environment_id} is protected. See environment setting: {data.PROTECTED_ENVIRONMENT}")
+                is_protected_environment = await env.get(data.PROTECTED_ENVIRONMENT, connection)
+                if is_protected_environment:
+                    raise Forbidden(
+                        f"Environment {environment_id} is protected. See environment setting: {data.PROTECTED_ENVIRONMENT}"
+                    )
+
+                # Check if the environment is halted; if not, halt it
+                if not env.halted:
+                    LOGGER.info("Halting Environment %s", str(environment_id))
+                    await self.halt(env, connection=connection)
```

Comment:
The `halt()` method doesn't have the `connection` argument yet, as far as I can see.